small/small.c

```diff
@@ -288,7 +288,8 @@ small_mempool_create(struct small_alloc *alloc)
 		 * at the next iteration. So we need to create the last
 		 * group of pools.
 		 */
-		if (objsize == alloc->objsize_max) {
+		if (objsize == alloc->objsize_max ||
+		    mempool_cache_size + 1 == SMALL_MEMPOOL_MAX) {
```

Comment:
The code is difficult to follow. Please rewrite so that the loop condition is checked in one place, for example, by moving the last group creation outside the loop or by using `break`.

Reply:
Done.

```diff
diff --git a/small/small.c b/small/small.c
index 1e6cfad..d82fd3f 100644
--- a/small/small.c
+++ b/small/small.c
@@ -281,20 +281,14 @@ small_mempool_create(struct small_alloc *alloc)
                                                    pool - 1);
                        cur_order_pool = pool;
                }
-               /*
-                * Maximum object size for mempool allocation ==
-                * alloc->objsize_max. If we have reached this size,
-                * there will be no more pools - loop will be broken
-                * at the next iteration. So we need to create the last
-                * group of pools.
-                */
-               if (objsize == alloc->objsize_max ||
-                   mempool_cache_size + 1 == SMALL_MEMPOOL_MAX) {
-                       assert(cur_order_pool->pool.slab_ptr_mask ==
-                              pool->pool.slab_ptr_mask);
-                       small_mempool_create_groups(alloc, cur_order_pool,
-                                                   pool);
-               }
+       }
+       if (alloc->small_mempool_cache_size > 0) {
+               int mempool_cache_size = alloc->small_mempool_cache_size;
+               struct small_mempool *last_pool =
+                       &alloc->small_mempool_cache[mempool_cache_size - 1];
+               assert(cur_order_pool->pool.slab_ptr_mask ==
+                      last_pool->pool.slab_ptr_mask);
+               small_mempool_create_groups(alloc, cur_order_pool, last_pool);
        }
        alloc->objsize_max = objsize;
 }
```